kernel/posix/mman.c

```diff
@@ -0,0 +1,598 @@
+ * how to close it. Ideally this would be in the k_thread, but this is a
+ * workaround.
+ */
+struct _fdte {
+	k_tid_t proc;
+	int fd;
+	int oflag;
+	struct _gfte *file;
+};
+
+/*
+ * Mappings table, tracking how a file descriptor was mapped and which pages it
+ * allocated.
+ */
+struct _mte {
+	k_tid_t proc;
+	struct _fdte *fd;
+	void *reg_start;
+	void *reg_end;
+	int map_type;
+	int prot;
+};
+
+static struct _gfte gft[GFT_ENTRIES];
+static struct _fdte fdt[FDT_ENTRIES];
+static struct _mte mt[MT_ENTRIES];
+
+/*
+ * Operations on global file table
+ */
+static struct _gfte *file_create(const char *name, mode_t perm);
+static int file_delete(struct _gfte *fe);
+static struct _gfte *file_find_by_name(const char *name);
+static int file_unlink(struct _gfte *fe);
+
+/*
+ * Operations on descriptors table
+ */
+static int fds_create(struct _gfte *gf, int oflags, struct _fdte **fde);
+static int fds_delete(struct _fdte *fde);
+static struct _fdte *fds_find_by_fd(int fd);
+static int fds_count_references(struct _gfte *fe);
+
+/*
+ * Operation on mappings table
+ */
+static struct _mte *map_create(struct _fdte *fde, int prot, int map_type);
+static int map_delete(struct _mte *me);
+static struct _mte *map_find_by_tid_addr(k_tid_t proc, void *addr);
+
+/*
+ * NULL if not found, pointer to the object if found.
+ */
+inline struct _gfte *file_find_by_name(const char *name)
+{
+	struct _gfte *it, *lmt, *fe = NULL;
+
+	if (name == NULL)
+		return NULL;
+
+	it = gft;
+	lmt = gft + GFT_ENTRIES;
+
+	while (it != lmt && !fe) {
+		if (!strncmp(name, it->name, PATH_MAX))
+			fe = it;
+		++it;
+	}
+
+	return fe;
+}
+
+/*
+ * No check for NULL
+ * Fails if no memory to allocate in the slab or GFT is full.
+ */
+inline struct _gfte *file_create(const char *name, mode_t perm)
+{
+	struct _gfte *it, *lmt, *fe = NULL;
+	void *data_ptr;
+
+	if (k_mem_slab_alloc(&mman_slab, &data_ptr, K_NO_WAIT))
+		return NULL;
+
+	it = gft;
+	lmt = gft + GFT_ENTRIES;
+
+	while (it != lmt && !fe) {
+		if (it->name == NULL && !it->to_unlink) {
+			fe = it;
+			fe->name = name;
+			fe->perm = perm;
+			fe->buff = data_ptr;
+			fe->to_unlink = false;
+		}
+		++it;
+	}
+	return fe;
+}
+
+inline int file_delete(struct _gfte *fe)
+{
+	if (!fe)
+		return -1;
+
+	k_mem_slab_free(&mman_slab, &fe->buff);
+	memset(fe, 0, sizeof(struct _gfte));
+
+	return 0;
+}
+
+inline int file_unlink(struct _gfte *fe)
+{
+	if (!fe)
+		return -1;
+
+	fe->name = NULL;
+	fe->to_unlink = true;
+	return 0;
+}
+
+/*
+ * TODO: consider that 0, 1, and 2 are reserved for stderr, stdin, stdout
+ */
+#define fds_valid(_fd) ((_fd) < FDS_PER_PROC)
+
+/*
+ * Needs to find an empty spot in the table, and still generate a valid fd,
+ * being the lowest possible.
+ */
+inline int fds_create(struct _gfte *gf, int oflags, struct _fdte **fde)
+{
+	struct _fdte *it, *lmt;
+	char used_fd[FDS_PER_PROC];
+	k_tid_t proc;
+	int fd = -1;
+
+	proc = k_current_get();
```

Comment:
Since k_current_get() is a syscall and proc isn't used until late in the function (and it can return early), it might be better to just call k_current_get() when attributing to (*fde)->proc.